### ISSUE TITLE
Microsite: Docusaurus 2 upgrade meta issue

### ISSUE THREAD

**Issue by jhaals** [backstage/backstage#12019]
Backstage.io currently runs on an old version of Docusaurus 1.0 which is no longer supported. We therefore want to upgrade to Docusaurus 2.2.

**The new microsite is available in the [microsite-next](https://github.com/backstage/backstage/tree/master/microsite-next) folder and can be started with `yarn start`**

The new microsite resides in the `microsite2` folder with symlinks to assets and other data from the old site. The plan is to move it to it's original location once the migration is complete.

This is a meta issue containing some of the outstanding work to address in order to switch out the microsite.

## Required
- [x] Add diagonal lines to startpage. This is what we had previously in the old microsite <img width="1118" alt="Screenshot 2023-02-22 at 13 47 41" src="https://user-images.githubusercontent.com/68980/220624171-776ae110-ddf5-42a1-852e-be2c9270604b.png">
- [x] Update plugin page category buttons to be more tag-like. <img width="814" alt="Screenshot 2023-02-22 at 13 49 49" src="https://user-images.githubusercontent.com/68980/220624396-920b065c-080b-4a73-b639-152367ae2760.png">
- [x] Migrate https://backstage.io/on-demand to the new microsite. This page is currently linked to from the [/community](https://backstage.io/community/) page.
- [x] Migrate https://backstage.io/nominate to the new microsite. This page is currently linked to from the [/community](https://backstage.io/community/) page.
- [x] https://backstage.io/docs/ does not take you to the correct place like the docs link in the header. #16593
- [x] Docsearch not working
- [x] Build configuration
The new plugin `src/plugins/backstage-plugin-list` currently need to be built before the microsite itself. 
This should be updated so that `yarn build` builds both the microsite and it's dependency in one go.
- [x] Forklift `microsite/src/core/components.js` to microsite-next. Minimal work to get the same look/feel for pages.
- [x] API References need to exist at build time
This breaks `yarn start` during development as `yarn build:api-docs` are typically not executed locally when developing. Ideally we should have some placeholder for this in the sidebar which is replaced when writing docs #15277
- [x] API references generated are not MDX compatible
`yarn build:api-docs` generate docs with HTML comments which breaks `yarn start` as these are not MDX compatible. We should investigate if these can be disabled, otherwise add pre start/build step that removes these comments using some regexp search similar to `<!--(.*?)-->` #15277
- [x] Too much spacing between plugin sections in the plugin marketplace
Needs slight css tweaks
- [x] Sidebar needs migration #15277
- [x] Update `<img>` tags in `docs/` to new [import format](https://docusaurus.io/docs/markdown-features/assets#files) (prior to switching site)
- [x] index page
- [x] /plugins marketplace. plugin marketplace(`/plugins`). Need to be converted to a Docusaurus plugin to support file loading as pages are plain React nowadays so we don't have access to `process` and similar.
- [x] /releases
- [x] /community
- [x] /demos

## Bonus

- [ ] Light mode. Looks great for generated content such as /docs but not for index pages which have hardcoded dark background. It's currently disabled in the docusaurus configuration file.

**Comment by chalin** [comment]
Hi @jhaals. In support of the microsite's migration to Google Analytics 4 (via #12658), a small config change is needed when migrating to Docusaurus v2:

```js
  presets: [
    [
      '@docusaurus/preset-classic',
      {
        docs: { ... },
        blog: { ... },
        theme: { ... },
        gtag: { // <= was googleAnalytics
          trackingID: 'UA-163836834-5',
        },
      },
    ],
  ],
```

[@docusaurus/plugin-google-gtag]: https://docusaurus.io/docs/api/plugins/@docusaurus/plugin-google-gtag

Actually, if possible, can you try setting up both the UA _and_ the GA4 tag? The GA4 tag is `G-KSEVGGNCJW`. If not, once the upgrade to Docusaurus v2 is complete, I can come back to experiment and make the necessary changes. Thanks!

**Comment by benjdlambert** [comment]
It's now GA. https://docusaurus.io/blog/2022/08/01/announcing-docusaurus-2.0

**Comment by jhaals** [comment]
@awanlin thanks for asking! I believe the majority of the roadblocks are those described in the issue description that all need to be tackled. You're more than welcome to help out with anyone of them that you feel comfortable with. This was mainly skunk work that I did before the summer and I'm not sure when we plan to pick it up again since the existing page if working but the experience could of course be better.

**Comment by marcofaggian** [comment]
Hi! I'd like to embarque on the journay of migrating the microsite to docusaurus:2.1, how can I help?

Should we take this issue as reference for the whole migration and break it out in smaller one for tracking single small improvements towards?

**Comment by jhaals** [comment]
Hi @marcofaggian, you are very welcome to help out! Some ground work started in `jhaals/microsite2` and it's probably the best base branch to work towards. I tried to capture some of the tasks some which are blocking us to move forward. Anything that takes us closer is welcome 🙏  I'm happy to add more things to this issue as we move forward.

**Comment by jhaals** [comment]
It's obvious that this is a fairly large chunk of work with many small dependencies. All of those dependencies make it really hard to gradually push this work forward as we have a moving main branch and my wip branch containing individual fixes that is going to lag behind unless someone updates it often.

The issues listed above need to be addressed at some point but it can be done in stages.
I have a suggestion on concrete initial steps for being able to gradually fix the microsite 🚀 

- [ ] Create a `microsite-next` directory **only** containing a fresh site with a CI job that ensures that the site is built when needed(and when `docs/` changes). Same as we do today basically.
- [ ] Make it display docs from `docs/`. This might require some compatibility/fixes as the docs need to be mdx compatible, but if we make sure it builds on both microsite and microsite-next is going to remain valid for both.
- [ ] Gradually add pages one by one
- [ ] Light mode 🙏 
- [ ] Celebrate 🎉

**Comment by jhaals** [comment]
microsite-next is now merged in so tasks can now be picked up in parallel.
The only thing that is wired up and working so far is /blog.

**Comment by waldirmontoya25** [comment]
Static assets in `/docs` apparently require to change the way they are linked. The one in the example would need to change to something like the following:

`<img
  src={require('../../docs/assets/utility-apis-fig1.png').default}
  alt="Figure showing the relationship between utility APIs, the apps that provide them, and the plugins that consume them"
/>`

This would work if the picture is changed from .svg to .png. I guess it would be good to standardize on .png given that most of the images are in that format.

**Comment by jhaals** [comment]
> Static assets in `/docs` apparently require to change the way they are linked. The one in the example would need to change to something like the following:
> 
> `<img src={require('../../docs/assets/utility-apis-fig1.png').default} alt="Figure showing the relationship between utility APIs, the apps that provide them, and the plugins that consume them" />`
> 
> This would work if the picture is changed from .svg to .png. I guess it would be good to standardize on .png given that most of the images are in that format.

Indeed looks like they have changed format for `<img>` html tags. We only use them in a few places where we've added additional styling such as centering the image. I'll update the bullet list to remember to change to the new img format prior to replacing the site, alternatively we could also live with a slightly uglier styling with

```diff
- <img src="../assets/utility-apis-fig1.svg" alt="Figure showing the relationship between utility APIs, the apps that provide them, and the plugins that consume them" />
+ ![Figure showing the relationship between utility APIs, the apps that provide them, and the plugins that consume them](./assets/utility-apis-fig1.svg)
```

Alternatively we might be able to style them with and have it work in both versions https://docusaurus.io/docs/markdown-features/assets#themed-images

**Comment by waldirmontoya25** [comment]
It looks like docusaurus v2 has a [theme](https://github.com/facebook/docusaurus/tree/main/packages/docusaurus-theme-classic/src/theme) with 100 components.

What would be the best way to approach the migration of `microsite/src/core/components.js` to `microsite-next` considering future upgrades? 

- [swizzling](https://docusaurus.io/docs/swizzling) existing theme components with custom ones?
- Create a new theme?
- Other options?

**Comment by jhaals** [comment]
> It looks like docusaurus v2 has a [theme](https://github.com/facebook/docusaurus/tree/main/packages/docusaurus-theme-classic/src/theme) with 100 components.
> 
> What would be the best way to approach the migration of `microsite/src/core/components.js` to `microsite-next` considering future upgrades?
> 
> * [swizzling](https://docusaurus.io/docs/swizzling) existing theme components with custom ones?
> * Create a new theme?
> * Other options?

Discussed this with the team and there's a risk that we end up with a lot of back end forth if we go for some replacement of the existing components which essentially is going to delay the migration. It's better to go with a full on forklift and do whatever necessary to get it to play well with docusaurus 2 by using `any` types for those components or simply keeping it `.js`

**Comment by luchillo17** [comment]
I'm starting to work on this, as far as I can see from latest master, stuff that doesn't work right now are:
- Header's logo doesn't show up, instead only the alt text does.
- Plugins, Demos & Community pages return the 404 page as content (not found content under the header).
- Home page is empty (it has "HELLO" flat as content)
- Docs work but look slightly different, are we sticking with that or try to adjust theme/layout to look like old site?
![Screenshot 2023-01-12 at 1 23 30 PM](https://user-images.githubusercontent.com/9698639/212148025-89bb16fc-fc28-45e3-ad5b-b4a9fd081995.png)
![Screenshot 2023-01-12 at 1 23 44 PM](https://user-images.githubusercontent.com/9698639/212148087-e158bd05-f29d-41fa-98a2-8de9d818957d.png)

**Comment by jhaals** [comment]
Hi @luchillo17, glad to hear that!
Yes the pages themselves including the home page does need to be migrated and there's likely tweaks that need to be made to how static assets are served as well for things such as logos in the header.
I'm not sure we should change the theme for the docs page unless the readability is better, I just want to make sure that we keep the builtin code blocks and similar looking good. My hope is that users will be able to switch between light/darkmode later on but we can force dark mode for now and tackle the other theme later.

**Comment by luchillo17** [comment]
So I can assume that the new theme/layout that Docusaurus 2 renders is ok for now?

**Comment by benjdlambert** [comment]
@luchillo17 yep indeed! Let's not overcomplicate the work required to get this working. Let's go with the bare minimum it's required to ship Docusaurus 2 :pray: think the new layout and theme for the docs looks pretty good anyways 👍

**Comment by luchillo17** [comment]
I see the 4 pages currently broken I mentioned were custom pages, no Docusaurus plugin, those were custom made, I was thinking, can we add MUI v5 to the Docusaurus site instead of creating handmade components? it would also help with consistency given Backstage will move eventually to MUI v5, @benjdlambert what do you think?

**Comment by Rugvip** [comment]
@luchillo17 I don't know whether we would want to add MUI v5 or introduce some more style frameworks in the future, but for the initial migration to Docusaurus 2.2 we should really aim to keep things simple. The existing pages are plain React/CSS and we should aim to lift those over with as few changes as possible. Once the migration is done we could take a look at making those pages a bit more friendly to work with, but the aim is to avoid a complete rewrite at this point.

**Comment by luchillo17** [comment]
Created a Draft with the **Community** page setup, a Draft for now since I want to tackle **Plugins** & **Demos** pages as well, would you prefer separate PRs?

**Comment by luchillo17** [comment]
Having some issues with Plugins page, old one was SSR so it would use NodeJS to read all YML files in the `microsite/data/plugins` folder, I don't think Docusaurus 2 can do the same, maybe import the YML files manually? or customizing the underlying Webpack to load those files properly? any suggestions?

**Comment by luchillo17** [comment]
I'll try `require.context` with a YML loader so Webpack process it for me.

**Comment by luchillo17** [comment]
Plugins page is working, styles are not pixel perfect but should be a good start.

**Comment by benjdlambert** [comment]
Nice work @luchillo17 - great to see this making progress!

**Comment by luchillo17** [comment]
Almost done, few things missing:

- Top left icon in page header, yet to fix
- Home page banner should be dismissible, yet to add the logic for that
- Home page has a `New Adopters` form popup fixed to the center-right position in the page, yet to add this
- Original theme was dark, so activating the light theme will have some of the sections look dark for now (those sections with color gradients), would like a designer to tell me what to do with such sections.

![Screenshot 2023-02-08 at 7 35 32 PM](https://user-images.githubusercontent.com/9698639/217684037-6778da1d-4e8a-4726-8e4d-d4dac6b9cc5f.png)
![Screenshot 2023-02-08 at 7 36 43 PM](https://user-images.githubusercontent.com/9698639/217684329-ec079067-747d-46d3-bfda-48ca644ec220.png)

**Comment by benjdlambert** [comment]
Looking great. One question I have is if it would be possible to still have the angles on the background instead of being flat lines like before? If it's too much effort then maybe we can skip it for now.

As for some of the design questions I'm going to try and get some feedback, but I wonder if we can either:

a) Force dark mode by default - maybe disable the light picker for now if thats possible?
b) Leave the dark sections as they are but make the text white instead for maybe a quick enough win?

**Comment by luchillo17** [comment]
I purposely ignored the angles, this PR is supposed to be a starting point, and since there's a need for a light theme I assumed we might get a new design that works with both themes.
Will search to see if dark mode can be forced, I can indeed force the text white in those sections if needed, though will leave that as a last resort for now.

**Comment by benjdlambert** [comment]
> I purposely ignored the angles, this PR is supposed to be a starting point

Sure, this makes sense.

> and since there's a need for a light theme I assumed we might get a new design that works with both themes.

Yep, I think you're right, its just to unblock this work so that we can start to ship bits maybe we can work around it for now and go with the dark theme for a start :)

**Comment by luchillo17** [comment]
The PR was merged and seems there's already more PRs building on top of it, what would be the next steps?

**Comment by jhaals** [comment]
@luchillo17 we have been working on getting it to build/deploy, speeding up the build and fixing search which we just go working again.

There's a couple of UI things that we would love to get help fixing.

One being the diagonal lines on the startpage. This is what we had previously.
<img width="1118" alt="Screenshot 2023-02-22 at 13 47 41" src="https://user-images.githubusercontent.com/68980/220624171-776ae110-ddf5-42a1-852e-be2c9270604b.png">

Secondly the categories on the plugin page are currently buttons and should probably be more tag-lookalike.
<img width="814" alt="Screenshot 2023-02-22 at 13 49 49" src="https://user-images.githubusercontent.com/68980/220624396-920b065c-080b-4a73-b639-152367ae2760.png">

**Comment by jhaals** [comment]
@luchillo17 I discovered that the /nominate and /on-demand are not migrated yet, I updated the issue to reflect that.
Secondly we noticed that https://backstage.io/docs does not take you to the same page as the docs header. I wonder if we can configure some sort of redirect there instead.

**Comment by luchillo17** [comment]
Those 2 routes were not mentioned before so I didn't look into them, as for the link, I think the links in the header can be configured from the config file, currently, I think it's pointing to `docs/overview/what-is-backstage`, we can change the link if you want, but adding a redirect to that link is another story, dunno if Docusaurus is capable of that.

I'll speak to the team to see if we create a new task that includes the diag line, and a custom tag component since Docusaurus components & Infima don't have components with a similar look, as well as the routing issues mentioned, would you prefer to change the header link to `/docs` or try and see if the redirect is possible with Docusaurus's router?

**Comment by benjdlambert** [comment]
> Also, I still think adding MUI is better than relying on the immature Infima GUI or making custom components...
@luchillo17 it looks like for now however we should be able to get away using the badges for the labels? https://infima.dev/docs/components/badge they look pretty OK and I think the `secondary` style or something should look ok. Adding MUI I feel is probably going to be a a step maybe in the wrong direction, as we don't have anything that looks in the way of MUI at all so that's going to be even more of an eyesore.

**Comment by luchillo17** [comment]
@benjdlambert Those badges didn't look close to the original so I didn't use them, indeed we can use Infima badges as tags, any chance we get a proper site design (Figma, Canva...) based on Infima components?

**Comment by jhaals** [comment]
> What about the question about the header link for docs? would you prefer the redirect or updating the header link to match https://backstage.io/docs?

I've added a redirect for /docs and pushed it out yesterday so that's taken care of with #16593. I'm not working on any of the other tasks just so you know.

**Comment by maciej-gurban** [comment]
Hi. I stumbled upon this issue trying to find info whether it would be possible to use Docusaurus to host the docs instead of MKDocs. We want to integrate solutions that we know work well in Docusaurus, e.g. live code demos , and we'd like to try to re-build using MKDocs, which seems more complex to achieve with additional learning curve for the plugin ecosystem.

At the same time, we'd like to be able to use the same solution across all our repos for documentation purposes so that we can index all docs content across our repos, and to my understanding to achieve that, that would require all docs to be using MKDocs/TechDocs.

Since your own site runs on Docusaurus, are you thinking of providing your customers the means of using Docusaurus or other markdown-based site generators? Looking at its popularity, Docusaurus [significantly surpasses MKDocs](https://www.wappalyzer.com/compare/mkdocs-vs-docusaurus/) in all popularity metrics, and subjectively, offers much nicer looking website with nearly no customization needed.

**Comment by Rugvip** [comment]
@maciej-gurban this issue is related to the microsite at https://backstage.io rather than TechDocs. There's been some discussion around supporting other SSGs in TechDocs though, see #8642

**Comment by luchillo17** [comment]
Our team has agreed on tasks for the 2 pages migration as well as the tags refactor, though not sure when I'll start on it yet.

**Comment by luchillo17** [comment]
PR for Nominate page: https://github.com/backstage/backstage/pull/16724

**Comment by luchillo17** [comment]
PR for diagonal lines for sections: #16888

**Comment by luchillo17** [comment]
Still see this open, no proper design yet?

**Comment by benjdlambert** [comment]
@luchillo17 nope - we're still waiting on a designer to come support us internally. Don't think there's a big rush for this though as yet, maybe we could close this ticket in favour of a smaller ticket for the other themes? Thoughts @jhaals?

**Comment by jhaals** [comment]
Created a new ticket specifically for light mode, closing this one 🙏
